dnf/transaction_sr.py

```diff
@@ -0,0 +1,169 @@
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Library General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+
+from __future__ import absolute_import
+from __future__ import print_function
+from __future__ import unicode_literals
+
+import libdnf
+
+from dnf.i18n import _
+import dnf.exceptions
+
+import json
+
+
+VERSION_MAJOR = "1"
+VERSION_MINOR = "0"
+VERSION = "%s.%s" % (VERSION_MAJOR, VERSION_MINOR)
+"""
+The version of the stored transaction.
+
+MAJOR version denotes backwards incompatible changes (old dnf won't work with
+new transaction JSON).
+
+MINOR version denotes extending the format without breaking backwards
+compatibility (old dnf can work with new transaction JSON), but breaks forwards
+compatibility (new dnf won't work with old transaction JSON).
+"""
+
+
+class TransactionError(dnf.exceptions.Error):
+    def __init__(self, msg):
+        super(TransactionError, self).__init__(msg)
+
+
+class TransactionFileError(dnf.exceptions.Error):
+    def __init__(self, filename, errors):
+        """
+        :param filename: The name of the transaction file being replayed
+        :param errors: a list of error classes or a string with an error description
+        """
+        if isinstance(errors, (list, tuple)):
+            if len(errors) > 1:
+                msg = _('Errors in "{filename}":').format(filename=filename)
+                for error in errors:
+                    msg += "\n  " + str(error)
+
+                super(TransactionFileError, self).__init__(msg)
+                return
+
+            else:
+                errors = str(errors[0])
+
+        msg = _('Error in "{filename}": {error}').format(filename=filename, error=errors)
+        super(TransactionFileError, self).__init__(msg)
+
+
+class IncompatibleTransactionVersionError(TransactionFileError):
+    def __init__(self, filename, msg):
+        super(IncompatibleTransactionVersionError, self).__init__(filename, msg)
+
+
+def check_version(version, filename):
+    major, minor = version.split('.')
+    if major != VERSION_MAJOR:
+        raise IncompatibleTransactionVersionError(
+            filename,
+            _('Incompatible major version {major}, supported major version is {major_supp}.')
+                .format(major=major, major_supp=VERSION_MAJOR)
+        )
+
+    if minor < VERSION_MINOR:
+        raise IncompatibleTransactionVersionError(
+            filename,
+            _('Incompatible minor version {minor}, '
+                'supported minor version is {minor_supp} or greater.'
+            ).format(minor=minor, minor_supp=VERSION_MINOR)
+        )
+
+
+def store_transaction(transaction, filename="transaction.json"):
+    data = {
+        "version": VERSION,
+    }
+    rpms = {}
```

Comment:
Please could you explain why you you want to use dict with NA keys? I see no advantage in this structure and only I see potential problems. What about to store each TSI as a separate object in the list. It will also help with replay, because it will reflect more precisely the original history records. In case that full transaction history record is stored in json, there will be no limitation in how the record can be used in the future (like for additional operation and not only replay - future commands or work-flows).

Reply:
We have discussed this before, but I'll repeat myself here.

I don't like the fact that some kinds of transaction items are split into two, e.g. upgrade `foo-1.0` -> `foo-2.0` is split into two transaction items: `Upgraded foo-1.0` and `Upgrade foo-2.0`, while in fact it is a single operation. I find it more concise and clear to store it as a single operation. Also there is data belonging to this single operation, which, if you split the operation into two, you either have to duplicate to store it with both transaction items, or pick one at random to store it with. E.g. `reason` that we store now with the transaction. Where would you put it if an upgrade is two separate transaction items? Neither option is good.

When storing data, it is important to store related data in a single place, not scatter them randomly across multiple records. In this case all data about an RPM package are stored under a single record, where you find all the operations that were performed with it. This also makes the format more future-proof, ff we want to store more data about an RPM or an action that was performed on it, there is a single correct place to put it.

I also find it easier to read and understand for anyone who'd like to inspect the file.

And, I see no real disadvantages to store the data this way. If you have anything concrete, please share it.

Reply:
Will your system work with packages that obsoletes packages without installing obsoletes?

Reply:
Sorry, I don't understand what you mean.

Reply:
I really like your approach to store data for each transaction element in dict, but I don't like the unification step where you put forward and reverse actions into one element. The unification that you want to use was part of dnf-2 design, but it was replaced in dnf-3/4 with SWDB, because the relations were incorrectly expressed. 

As you mentioned you like your approach, but that the only reason why you want to implement it that way. or did I overlooked something?

I used several examples where your approach was failing. You improved it to not fail in those particular examples but not more. And I am also not happy about how you resolved issues => structure that could be dict or list of dicts. 

Last time I pointed out that the design will fail when multiple packages with the same name-arch will be obsoleted.

Also similar problems could appear with groups/env-groups - in transaction could be multiple actions with the same group ID, you your design will only store the last action => random results.

I recommended you to replace small but an important part of your design - replace Name-Arch dict with dict or list of dicts by a list with dicts. Then there will be no issue with any rare cases in transaction, because json file will mirror the history transaction. Then every user of the json file will decide how to use stored data.

Reply:
> I really like your approach to store data for each transaction element in dict, but I don't like the unification step where you put forward and reverse actions into one element. The unification that you want to use was part of dnf-2 design, but it was replaced in dnf-3/4 with SWDB, because the relations were incorrectly expressed.

This is once again just unhelpful comparison. I have no idea what the design in dnf-2 was and what was the issue with it. If you know of a concrete problem with this representation, I would very much like to know it, please! So far you have only told me there was a problem, or that there could be a problem, but you have never told me what the problem was/is.

> As you mentioned you like your approach, but that the only reason why you want to implement it that way. or did I overlooked something?

I've summarized the reasons in https://github.com/rpm-software-management/dnf/pull/1630#discussion_r444207656 and I consider them valid arguments for storing the data the way I did.

> I used several examples where your approach was failing. You improved it to not fail in those particular examples but not more. ...

What more are you asking for? The only issue AFAIR were `installonly` packages. The support for them was just missing from the initial implementation. I know of no other issues.

> ... And I am also not happy about how you resolved issues => structure that could be dict or list of dicts.

I can understand that, but (and I am repeating myself here) I find it a different issue to the actual format of storing transaction items. This can be solved by simply always using an array (with a single item for non-`installonly` packages). But, as I said, I think it's fine the way it is.

> Last time I pointed out that the design will fail when multiple packages with the same name-arch will be obsoleted.

Sure, and I have no idea what you meant by that, so I asked you for an explanation, which you never provided.

> Also similar problems could appear with groups/env-groups - in transaction could be multiple actions with the same group ID, you your design will only store the last action => random results.

I've replied to this already as well. It doesn't make any sense to have a transaction with multiple operations on a single group id (unless I'm missing something, in that case, again, please explain). There are rules to how dnf works and the rules define how the transaction data is ultimately stored.

> I recommended you to replace small but an important part of your design - replace Name-Arch dict with dict or list of dicts by a list with dicts. Then there will be no issue with any rare cases in transaction, because json file will mirror the history transaction. Then every user of the json file will decide how to use stored data.

Once again, I don't really understand what you are suggesting. Please provide example JSON snippets. (But I still don't see any justification to change the format)

Reply:
> >   Also similar problems could appear with groups/env-groups - in transaction could be multiple actions with the same group ID, you your design will only store the last action => random results.

> I've replied to this already as well. It doesn't make any sense to have a transaction with multiple operations on a single group id (unless I'm missing something, in that case, again, please explain). There are rules to how dnf works and the rules define how the transaction data is ultimately stored.

The main point is that it could appear in transaction and your code is not ready for that situation. I really don't care how it could happened or how the situation is rare. In case that you want additional information and to know what kind of problems could be in transaction please open Bugzilla and check system-upgrade issues. Then you will see that everything is possible. I also would like to mention that your code is going to be used by system-upgrade plugin, therefore it must be bullet-proof even against rare or impossible situations.